gcc/rust/lex/rust-lex.h

```diff
@@ -232,11 +232,128 @@ class Lexer
   // Input source wrapper thing.
   class InputSource
   {
+  private:
+    unsigned int pos;
+    std::vector<Codepoint> chars;
+
+    // Overload operator () to return next char from input stream.
+    virtual int next_byte () = 0;
+
+    Codepoint next_codepoint ()
+    {
+      uint8_t input = next_byte ();
+
+      if ((int8_t) input == EOF)
+	return Codepoint::eof ();
+      // check for UTF-8 BOM
+      if (pos == 0 && input == 0xEF)
+	{
+	  uint8_t input2 = next_byte ();
+	  if (input2 != 0xBB)
+	    return {0xFFFE};
+	  uint8_t input3 = next_byte ();
+	  if (input3 != 0xBF)
+	    return {0xFFFE};
+	  // UTF-8 BOM is found
+	  return next_codepoint ();
+	}
+      else if (input < 128)
+	{
+	  // ascii -- 1 byte
+	  return {input};
+	}
+      else if ((input & 0xC0) == 0x80)
+	{
+	  // invalid (continuation; can't be first char)
+	  return {0xFFFE};
+	}
+      else if ((input & 0xE0) == 0xC0)
+	{
+	  // 2 bytes
+	  uint8_t input2 = next_byte ();
+	  if ((input2 & 0xC0) != 0x80)
+	    return {0xFFFE};
+
+	  uint32_t output = ((input & 0x1F) << 6) | ((input2 & 0x3F) << 0);
+	  return output;
+	}
+      else if ((input & 0xF0) == 0xE0)
+	{
+	  // 3 bytes
+	  uint8_t input2 = next_byte ();
+	  if ((input2 & 0xC0) != 0x80)
+	    return {0xFFFE};
+
+	  uint8_t input3 = next_byte ();
+	  if ((input3 & 0xC0) != 0x80)
+	    return {0xFFFE};
+
+	  uint32_t output = ((input & 0x0F) << 12) | ((input2 & 0x3F) << 6)
+			    | ((input3 & 0x3F) << 0);
+	  return {output};
+	}
+      else if ((input & 0xF8) == 0xF0)
+	{
+	  // 4 bytes
+	  uint8_t input2 = next_byte ();
+	  if ((input2 & 0xC0) != 0x80)
+	    return {0xFFFE};
+
+	  uint8_t input3 = next_byte ();
+	  if ((input3 & 0xC0) != 0x80)
+	    return {0xFFFE};
+
+	  uint8_t input4 = next_byte ();
+	  if ((input4 & 0xC0) != 0x80)
+	    return {0xFFFE};
+
+	  uint32_t output = ((input & 0x07) << 18) | ((input2 & 0x3F) << 12)
+			    | ((input3 & 0x3F) << 6) | ((input4 & 0x3F) << 0);
+	  return {output};
+	}
+      else
+	{
+	  // rust_error_at (get_current_location (),
+	  //   "invalid UTF-8 [SECND] (too long)");
+	  return {0xFFFE};
+	}
+    }
+
   public:
+    InputSource () : pos (0), chars ({}) {}
+
     virtual ~InputSource () {}
 
-    // Overload operator () to return next char from input stream.
-    virtual int next () = 0;
+    // Check if the input source is valid as utf-8 and copy all characters to
+    // `chars`.
+    void init ()
+    {
```

Comment:
modified `InputSource` to check the input string is valid utf-8 and push utf-8 characters to its buffer immidiately after an instance of this class is created.